features/cellular/framework/AT/ATHandler.cpp

```diff
@@ -1287,30 +1284,45 @@ void ATHandler::flush()
     }
 }
 
-void ATHandler::debug_print(const char *p, int len)
+void ATHandler::debug_print(const char *p, int len, ATType type)
 {
 #if MBED_CONF_CELLULAR_DEBUG_AT
     if (_debug_on) {
-#if MBED_CONF_MBED_TRACE_ENABLE
-        mbed_cellular_trace::mutex_wait();
-#endif
-        char c;
-        for (ssize_t i = 0; i < len; i++) {
-            c = *p++;
-            if (!isprint(c)) {
-                if (c == '\r') {
-                    debug("\n");
+        const int buf_size = len * 4 + 1; // x4 -> reserve space for extra characters, +1 -> terminating null
+        char *buffer = (char *)malloc(buf_size);
+        if (buffer) {
+            memset(buffer, 0, buf_size);
+
+            char *pbuf = buffer;
+            for (ssize_t i = 0; i < len; i++) {
+                const char c = *p++;
+                if (isprint(c)) {
+                    *pbuf++ = c;
+                } else if (c == '\r') {
+                    sprintf(pbuf, "<cr>");
+                    pbuf += 4;
                 } else if (c == '\n') {
+                    sprintf(pbuf, "<ln>");
+                    pbuf += 4;
                 } else {
-                    debug("#%02x", c);
+                    sprintf(pbuf, "<%02X>", c);
+                    pbuf += 4;
                 }
+            }
+            MBED_ASSERT((int)(pbuf - buffer) <= buf_size); // Check for buffer overflow
+
+            if (type == AT_RX) {
+                tr_info("AT RX (%2d): %s", len, buffer);
+            } else if (type == AT_TX) {
+                tr_info("AT TX (%2d): %s", len, buffer);
             } else {
-                debug("%c", c);
+                tr_info("AT ERR (%2d): %s", len, buffer);
             }
+
+            delete buffer;
```

Comment:
You get the buffer from `malloc()` and free it by calling `delete`. This is a bit inconstant.

Use `malloc()/free()` pair or `new/delete` pair.

Reply:
Good catch. Fixed.